lib/puppet/provider/mongodb_replset/mongo.rb

```diff
@@ -201,12 +232,33 @@ def set_members
       if output['ok'] == 0
         raise Puppet::Error, "rs.initiate() failed for replicaset #{self.name}: #{output['errmsg']}"
       end
+
+      # Check that the replicaset has finished initialization
+      retry_limit = 10
+      retry_sleep = 3
+
+      retry_limit.times do |n|
+        begin
+          if db_ismaster(alive_hosts[0])['ismaster']
+            Puppet.debug 'Replica set initialization has successfully ended'
+            return
+          else
+            Puppet.debug "Wainting for replica initialization. Retry: #{n}"
+            sleep retry_sleep
+            next
+          end
+        end
+      end
+      raise Puppet::Error, "rs.initiate() failed for replicaset #{self.name}: host #{alive_hosts[0]} didn't become master"
+
     else
       # Add members to an existing replset
+      Puppet.debug "Adding member to existing replset #{self.name}"
       if master = master_host(alive_hosts)
         current_hosts = db_ismaster(master)['hosts']
```

Comment:
@timogoebel Sounds sensible... However I've not been able to replicate this repeat addition... 

Are you able to gist a debug log? 

Cheers